### DISCUSSION TITLE
运行basic_handshaking_test的时候，能正常收到cmd指令和发出报文，但是不能接收到任何报文

### DISCUSSION THREAD

**Discussion by lhyyy1** [SUSTech-CS305-Fall22/CS305-Project-Skeleton#45] [Q&A]
如图所示，在这个地方抛出一个异常
![image](https://user-images.githubusercontent.com/120730404/210052887-a5fa3b08-9abf-40bb-8d25-969bf89bd5c5.png)
七个peer中只有一个抛出了异常，说明只有收到cmd指令的那个peer读到了东西。
如图所示，在这个地方抛出异常
![image](https://user-images.githubusercontent.com/120730404/210053172-ca5c7c84-7b47-4ff7-b627-9509f632e6f6.png)
说明报文是可以正常发出的（实际上WHOHAS的那个点也过了）
我基本没有改动process_download和process_inbound_udp之外的任何代码，除了定义了几个全局变量（也就是说，和dumbsender和dumbreceiver是一样的）
![image](https://user-images.githubusercontent.com/120730404/210053330-8a708663-9643-4e99-bd01-62fa7e1d50eb.png)

**Comment by TANSixu** [comment]
Hi,方便提供一下测试失败后的log信息嘛，peer1.log, checker.log, peer2.log

**Reply by lhyyy1** [reply]
![image](https://user-images.githubusercontent.com/120730404/210058731-6fe27087-24be-4528-96ec-e4f34b37f8cf.png)

**Reply by lhyyy1** [reply]
如果我抛出异常的话log会变成这样（
![image](https://user-images.githubusercontent.com/120730404/210059499-bcee19d9-4b18-4232-b052-75d2e75a26d0.png)

**Reply by TANSixu** [reply]
Hi, 这应该是peer没有完全启动，请修改以下代码：
在grader.py中，class peerproc：
```
    def start_peer(self):
        if self.timeout:
            cmd = f"python3 -u {self.peer_file_loc} -p {self.node_map_loc} -c {self.haschunk_loc} -m {self.max_transmit} -i {self.id} -t {self.timeout}"
        else:
            cmd = f"python3 -u {self.peer_file_loc} -p {self.node_map_loc} -c {self.haschunk_loc} -m {self.max_transmit} -i {self.id}"

        self.process = subprocess.Popen(cmd.split(" "), stdin=subprocess.PIPE,stdout=subprocess.DEVNULL,text=True, bufsize=1, universal_newlines=True)
        # ensure peer is running
        time.sleep(0.1) 
```
将sleep0.1改成1，确保peer完全启动。
本周也有其他同学遇到了类似情况，多是由于import了更多包导致启动变慢。目前1s的启动时间暂时没有出过问题，我们将会在下次放test更新时将此处正式更新为1s。

**Reply by lhyyy1** [reply]
我这样操作了，但还是发生了完全一样的情况
即使sleep10秒也是一样的情况（

**Reply by TANSixu** [reply]
> 我这样操作了，但还是发生了完全一样的情况 即使sleep10秒也是一样的情况（

hi同学，请联系一下我qq（846030067），咱们再试试其他方法哈

**Comment by lhyyy1** [comment]
感谢学长耐心地帮我debug，这个问题的解答：peer.py抛出的异常不一定能看到，可能是peer.py的某个地方报错了。我自己的情况就是peer.py有个地方出了个语法错误
